xonsh/environ.py

```diff
@@ -42,6 +43,24 @@
 import xonsh.prompt.base as prompt
 
 
+events.doc('on_envvar_new', """
+on_envvar_new(name: str, value: Any) -> None
+
+Fires after a new enviroment variable is created.
+Warning! if a handler tries to set an envvar
+you'll get a nice recursion until the limit.
+""")
+
+
+events.doc('on_envvar_change', """
+on_envvar_change(name: str, oldvalue: Any, newvalue: Any) -> None
+
+Fires after an enviroment variable is changed.
+Warning! if a handler tries to set an envvar
```

Comment:
I would prefer a less hopeless warning, since it'll only recurse if they make unconditional changes.